File: app/src/main/java/com/matchr/LoginActivity.java
package com.matchr;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.Toast;

import com.google.android.gms.auth.api.Auth;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.auth.api.signin.GoogleSignInResult;
import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.common.api.ResultCallback;
import com.google.android.gms.common.api.Status;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.matchr.activities.QuestionActivity;
import com.matchr.data.Response;
import com.matchr.utils.L;

import java.util.ArrayList;

/**
 * A login screen that offers login via email/password.
 */
public class LoginActivity extends AppCompatActivity implements View.OnClickListener, GoogleApiClient.OnConnectionFailedListener {

    private static final int REQ_CODE = 9001;
    private GoogleApiClient googleApiClient;
    FirebaseDatabase database = FirebaseDatabase.getInstance();
    DatabaseReference myRef = database.getReference("users");

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        findViewById(R.id.signin_button).setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                signIn();

            }
        });

        findViewById(R.id.bn_logout).setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                signOut();

            }
        });

        GoogleSignInOptions signInOptions = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN).requestEmail().build();
        googleApiClient = new GoogleApiClient.Builder(this).enableAutoManage(this, this).addApi(Auth.GOOGLE_SIGN_IN_API, signInOptions).build();
        ArrayList<String> test = new ArrayList<>();
        test.add("hi");
        test.add("bye");
        Firebase.INSTANCE.saveResponse("abcd", new Response(25, test));
        Firebase.INSTANCE.test();
        signIn();
    }

    private void signIn() {
        Intent intent = Auth.GoogleSignInApi.getSignInIntent(googleApiClient);
        startActivityForResult(intent, REQ_CODE);

    }

    private void signOut() {
        Auth.GoogleSignInApi.signOut(googleApiClient).setResultCallback(
                new ResultCallback<Status>() {
                    @Override
                    public void onResult(Status status) {
                        Context context = getApplicationContext();
                        Toast.makeText(context, "Logged out", Toast.LENGTH_SHORT).show();
                    }
                });
    }

    private void handleResult(GoogleSignInResult result) {
        // if successfully signed in with google
        if (googleApiClient != null && googleApiClient.isConnected()) {
            GoogleSignInAccount account = result.getSignInAccount();
            if (account == null) {
                L.INSTANCE.d("Account is null", null);
                return;
            }
            // contains user ID
            String personId = account.getId();
            String personName = account.getDisplayName();
            String personEmail = account.getEmail();
            String givenName = account.getGivenName();
            String comb = "u" + givenName;

            myRef.child(comb + "/UserId").setValue(personId);
            myRef.child(comb + "/Name").setValue(personName);
            myRef.child(comb + "/Email").setValue(personEmail);

            // LoginActivity.class will be changed with next activity
            Intent intent = new Intent(this, QuestionActivity.class);
            intent.putExtra(Firebase.USER_ID, personId);
            startActivity(intent);
        }
    }

    @Override
    public void onConnectionFailed(@NonNull ConnectionResult connectionResult) {

    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.bn_logout:
                signOut();
                break;
            case R.id.signin_button:
                signIn();
                break;
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == REQ_CODE) {
            GoogleSignInResult result = Auth.GoogleSignInApi.getSignInResultFromIntent(data);
            handleResult(result);
        }
    }
}

File: app/src/main/java/com/matchr/LoginActivity.kt
package com.matchr

import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.widget.Toast
import com.google.android.gms.auth.api.Auth
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.auth.api.signin.GoogleSignInResult
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.api.GoogleApiClient
import com.matchr.activities.QuestionActivity
import com.matchr.data.User
import com.matchr.utils.L

/**
 * A login screen that offers login via email/password.
 */
class LoginActivity : AppCompatActivity(), View.OnClickListener, GoogleApiClient.OnConnectionFailedListener {
    private var googleApiClient: GoogleApiClient? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        findViewById<View>(R.id.signin_button).setOnClickListener { signIn() }

        findViewById<View>(R.id.bn_logout).setOnClickListener { signOut() }

        val signInOptions = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN).requestEmail().build()
        googleApiClient = GoogleApiClient.Builder(this).enableAutoManage(this, this).addApi(Auth.GOOGLE_SIGN_IN_API, signInOptions).build()
        if (BuildConfig.DEBUG) signIn()
    }

    private fun signIn() {
        val intent = Auth.GoogleSignInApi.getSignInIntent(googleApiClient)
        startActivityForResult(intent, REQ_CODE)

    }

    private fun signOut() {
        Auth.GoogleSignInApi.signOut(googleApiClient).setResultCallback {
            val context = applicationContext
            Toast.makeText(context, "Logged out", Toast.LENGTH_SHORT).show()
        }
    }

    private fun handleResult(result: GoogleSignInResult) {
        // if successfully signed in with google
        if (googleApiClient != null && googleApiClient!!.isConnected) {
            val account = result.signInAccount
            if (account == null) {
                L.d("Account is null", null)
                return
            }
            // contains user ID
            val id = account.id ?: return
            val name = account.displayName ?: "Unnamed"
            val email = account.email ?: "NA"
            Firebase.saveUser(User(id, name, email))

            // LoginActivity.class will be changed with next activity
            val intent = Intent(this, QuestionActivity::class.java)
            //            intent.putExtra(Firebase.USER_ID, "u" + System.currentTimeMillis());
            intent.putExtra(Firebase.USER_ID, id)
            if (BuildConfig.DEBUG) Firebase.test(id)
            startActivity(intent)
        }
    }

    override fun onConnectionFailed(connectionResult: ConnectionResult) {

    }

    override fun onClick(view: View) {
        when (view.id) {
            R.id.bn_logout -> signOut()
            R.id.signin_button -> signIn()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == REQ_CODE) {
            val result = Auth.GoogleSignInApi.getSignInResultFromIntent(data)
            handleResult(result)
        }
    }

    companion object {

        private val REQ_CODE = 9001
    }
}